### PR TITLE
coord: unify timestamp determination for peeks and tails

### PR DESCRIPTION
**Pull request by benesch**
Timestamp determination for peeks and tails has converged. #11037
allowed peeking unmaterialized sources; #10997 removed the "no complete
timestamps" error. So there is no longer any reason to have separate
timestamp determination paths.

This commit removes the `determine_frontier` method. Tails now use the
same `determine_timestamp` method as peeks. I believe this is roughly
behavior preserving, but I didn't spend the time to fully understand the
intracacies of `determine_frontier`.

This commit also changes `CREATE SINK` to choose the earliest possible
as of; essentially sink as ofs no longer take linearizability into
account. I belive this to be good policy for the reasons described in
the comment within the patch.

### Motivation

   * This PR refactors existing code.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by frankmcsherry** [comment]
> This commit also changes CREATE SINK to choose the earliest possible
as of;

Slightly weird to see this, and also see all the TAIL timestamps change from `Timestamp::MIN` to `Timestamp::MAX`. That might be surprising?

**Review by frankmcsherry — COMMENTED** [review]
I like this, except for the part where TAIL now produces outputs that are at the end of time. That seems problematic for anyone using TAIL for application use cases where they want to know when a value comes into effect ("not yet" for all of those TAIL queries, even though we know it should).

I don't know if this should block merging this, but I'll start reading to see if it is an easy fix.

File: src/coord/src/coord.rs [hunk 1]
```diff
@@ -3039,27 +3046,18 @@ impl Coordinator {
...
-                    compute_instance,
-                )?;
-                Antichain::from_elem(ts)
-            } else {
-                coord.determine_frontier(uses, compute_instance)
-            };
+            let id_bundle = coord.catalog.nearest_indexes(uses);
+            // If a timestamp was explicitly requested, use that.
+            let timestamp =
+                coord.determine_timestamp(session, &id_bundle, when, compute_instance)?;
```
Comment (frankmcsherry) [inline src/coord/src/coord.rs:3052]: I think I might prefer this being 'coord.least_valid_read()` instead? I guess there is a legit question about whether TAIL should present as much or as little historical context that it has. Counter-argument: commands that do not have `AS OF` in them should present as sequentially ordered. It's a pity choosing such a late time for e.g. constant queries though.

**Review by mjibson — COMMENTED** [review]
LGTM

**Pull request merged**